pkg/manager/member/tiproxy_member_manager.go

```diff
@@ -0,0 +1,656 @@
+			"startup-script": startScript,
+		},
+	}
+
+	var inUseName string
+	if set != nil {
+		inUseName = mngerutils.FindConfigMapVolume(&set.Spec.Template.Spec, func(name string) bool {
+			return strings.HasPrefix(name, controller.TiProxyMemberName(tc.Name))
+		})
+	}
+
+	klog.Info("get tiproxy in use config map name: ", inUseName)
+
+	err = mngerutils.UpdateConfigMapIfNeed(m.deps.ConfigMapLister, tc.BaseTiProxySpec().ConfigUpdateStrategy(), inUseName, newCm)
+	if err != nil {
+		return nil, err
+	}
+
+	return m.deps.TypedControl.CreateOrUpdateConfigMap(tc, newCm)
+}
+
+func (m *tiproxyMemberManager) syncStatefulSet(tc *v1alpha1.TidbCluster) error {
+	ns := tc.GetNamespace()
+	tcName := tc.GetName()
+
+	oldStsTmp, err := m.deps.StatefulSetLister.StatefulSets(ns).Get(controller.TiProxyMemberName(tcName))
+	if err != nil && !errors.IsNotFound(err) {
+		return fmt.Errorf("syncStatefulSet: failed to get sts %s for cluster %s/%s, error: %s", controller.TiProxyMemberName(tcName), ns, tcName, err)
+	}
+
+	stsNotExist := errors.IsNotFound(err)
+	oldStatefulSet := oldStsTmp.DeepCopy()
+
+	// failed to sync tiproxy status will not affect subsequent logic, just print the errors.
+	if err := m.syncStatus(tc, oldStatefulSet); err != nil {
+		klog.Errorf("failed to sync TidbCluster: [%s/%s]'s tiproxy status, error: %v",
+			ns, tcName, err)
+	}
+
+	cm, err := m.syncConfigMap(tc, oldStatefulSet)
+	if err != nil {
+		return err
+	}
+
+	newSts, err := m.getNewStatefulSet(tc, cm)
+	if err != nil {
+		return err
+	}
+
+	if stsNotExist {
+		err = mngerutils.SetStatefulSetLastAppliedConfigAnnotation(newSts)
+		if err != nil {
+			return err
+		}
+		err = m.deps.StatefulSetControl.CreateStatefulSet(tc, newSts)
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+
+	// Scaling takes precedence over upgrading because:
+	// - if a pod fails in the upgrading, users may want to delete it or add
+	//   new replicas
+	// - it's ok to scale in the middle of upgrading (in statefulset controller
+	//   scaling takes precedence over upgrading too)
+	if err := m.scaler.Scale(tc, oldStatefulSet, newSts); err != nil {
+		return err
+	}
+
+	if !templateEqual(newSts, oldStatefulSet) || tc.Status.TiProxy.Phase == v1alpha1.UpgradePhase {
+		if err := m.upgrader.Upgrade(tc, oldStatefulSet, newSts); err != nil {
+			return err
+		}
+	}
+
+	return mngerutils.UpdateStatefulSetWithPrecheck(m.deps, tc, "FailedUpdateTiProxySTS", newSts, oldStatefulSet)
+}
+
+func (m *tiproxyMemberManager) syncStatus(tc *v1alpha1.TidbCluster, sts *apps.StatefulSet) error {
+	if sts == nil {
+		// skip if not created yet
+		return nil
+	}
+
+	tc.Status.TiProxy.StatefulSet = &sts.Status
+	upgrading, err := m.statefulSetIsUpgradingFn(m.deps.PodLister, m.deps.PDControl, sts, tc)
+	if err != nil {
+		tc.Status.TiProxy.Synced = false
+		return err
+	}
+	if upgrading {
+		tc.Status.TiProxy.Phase = v1alpha1.UpgradePhase
+	} else {
+		tc.Status.TiProxy.Phase = v1alpha1.NormalPhase
+	}
+
```

Comment:
Is `tc.Status.PD.Phase = v1alpha1.ScalePhase` needed here?

Reply:
It is changed.